src/server/modules/threat/services/SafeBrowsingService.ts

```diff
@@ -89,6 +90,76 @@ export class SafeBrowsingService implements UrlThreatScanService {
     }
     return matches
   }
+
+  // TODO: deduplicate this function with the lookup function above
+  // ideally, we should do a non-block write for any matched viruses
+  // into the redis cache. the write above is blocking so we chose to
+  // bypass the write for now
+  private async lookupBulk(urls: string[]) {
+    let matches = null
+    const request = { ...this.requestTemplate } as any
+
+    request.threatInfo.threatEntries = urls.map((url) => {
+      return { url }
+    })
+
+    const response = await fetch(ENDPOINT, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(request),
+    })
+    if (!response.ok) {
+      const error = new Error(
+        `Safe Browsing failure:\tError: ${response.statusText}\thttpResponse: ${response}\t body:${response.body}`,
+      )
+      if (safeBrowsingLogOnly) {
+        logger.error(error)
+      } else {
+        throw error
+      }
+    } else {
+      const result = await response.json()
+      if (result?.matches) {
+        const prefix = safeBrowsingLogOnly
+          ? 'Considered threat by Safe Browsing but ignoring'
+          : 'Malicious link content'
+        result.matches.forEach((threatMatch: any) => {
+          logger.warn(
+            `${prefix}: ${threatMatch.threat.url} yields ${JSON.stringify(
+              threatMatch,
+              null,
+              2,
+            )}`,
+          )
+        })
+        matches = result.matches
+      }
+    }
+    return matches
+  }
+
+  public isThreatBulk: (urls: string[]) => Promise<boolean> = async (urls) => {
+    const urlChunks: string[][] = []
+
+    for (let i = 0; i < urls.length; i += SAFE_BROWSING_LIMIT) {
+      urlChunks.push(urls.slice(i, i + SAFE_BROWSING_LIMIT))
+    }
+
+    if (!safeBrowsingKey) {
+      logger.warn(`No Safe Browsing API key provided. Not scanning in bulk`)
+      return false
+    }
+
+    for (let i = 0; i < urlChunks.length; i += 1) {
+      // eslint-disable-next-line no-await-in-loop
+      const match = await this.lookupBulk(urlChunks[i])
+      if (!safeBrowsingLogOnly && Boolean(match)) {
```

Comment:
sidenote: don't see the environment variable `SAFE_BROWSING_LOG_ONLY` (original PR #376) being defined in the readme, should put it in there sometime

Reply:
opened as separate issue